motoko/invoice-canister/src/invoice/main.mo

```diff
@@ -0,0 +1,450 @@
+  type Details = T.Details;
+  type Token = T.Token;
+  type TokenVerbose = T.TokenVerbose;
+  type AccountIdentifier = T.AccountIdentifier;
+  type Invoice = T.Invoice;
+// #endregion
+
+  let errInvalidToken =
+    #err({
+       message = ?"This token is not yet supported. Currently, this canister supports ICP.";
+       kind = #InvalidToken;
+    });
+
+/**
+* Application State
+*/
+
+// #region State
+  stable var entries : [(Nat, Invoice)] = [];
+  stable var invoiceCounter : Nat = 0;
+  let invoices : HashMap.HashMap<Nat, Invoice> = HashMap.fromIter(Iter.fromArray(entries), entries.size(), Nat.equal, Hash.hash);
+  entries := [];
+  let MAX_INVOICES = 30_000;
+// #endregion
+
+/**
+* Application Interface
+*/
+
+// #region Create Invoice
+  public shared ({caller}) func create_invoice (args : T.CreateInvoiceArgs) : async T.CreateInvoiceResult {
+    let id : Nat = invoiceCounter;
+    // increment counter
+    invoiceCounter += 1;
+    let inputsValid = areInputsValid(args);
+    if(not inputsValid) {
+      return #err({
+        message = ?"Bad size: one or more of your inputs exceeds the allowed size.";
+        kind = #BadSize;
+      });
+    };
+
+    if(id > MAX_INVOICES){
+      return #err({
+        message = ?"The maximum number of invoices has been reached.";
+        kind = #MaxInvoicesReached;
+      });
+    };
+
+    let destinationResult : T.GetDestinationAccountIdentifierResult = getDestinationAccountIdentifier({
+      token = args.token;
+      invoiceId = id;
+      caller
+    });
+
+    switch(destinationResult){
+      case (#err result) {
+        #err({
+          message = ?"Invalid destination account identifier";
+          kind = #InvalidDestination;
+        });
+      };
+      case (#ok result) {
+        let destination : AccountIdentifier = result.accountIdentifier;
+        let token = getTokenVerbose(args.token);
+
+        let invoice : Invoice = {
+          id;
+          creator = caller;
+          details = args.details;
+          permissions = args.permissions;
+          amount = args.amount;
+          amountPaid = 0;
+          token;
+          verifiedAtTime = null;
+          paid = false;
+          destination;
+        };
+
+        invoices.put(id, invoice);
+
+        #ok({invoice});
+      };
+    };
+  };
+
+  func getTokenVerbose(token : Token) : TokenVerbose {
+    switch(token.symbol){
+      case ("ICP") {
+        {
+          symbol = "ICP";
+          decimals = 8;
+          meta = ?{
+            Issuer = "e8s";
+          }
+        };
+
+      };
+      case (_) {
+        {
+          symbol = "";
+          decimals = 1;
+          meta = ?{
+            Issuer = "";
+          }
+        }
+      };
+    };
+  };
+
+  func areInputsValid(args : T.CreateInvoiceArgs) : Bool {
+    let token = getTokenVerbose(args.token);
+
+    var isValid = true;
+
+    switch (args.details){
+      case null {};
+      case (? details){
+        if (details.meta.size() > 32_000) {
```

Comment:
Suggestion: define these magic numbers (32000, 256) as constants in some central location

Reply:
I'll create a follow-up issue for this